probe-rs/src/probe/blackmagic/mod.rs

```diff
@@ -0,0 +1,1508 @@
+            Self::Unsupported(e) => write!(f, "Remote command unsupported with result {:016x}", *e),
+            Self::ProbeError(e) => write!(f, "Probe error {}", e),
+        }
+    }
+}
+
+impl ProbeError for RemoteError {}
+
+struct RemoteResponse(u64);
+
+#[derive(PartialEq, Copy, Clone)]
+enum SwdDirection {
+    Input,
+    Output,
+}
+
+impl From<bool> for SwdDirection {
+    fn from(value: bool) -> Self {
+        if value {
+            SwdDirection::Output
+        } else {
+            SwdDirection::Input
+        }
+    }
+}
+
+/// A Black Magic Probe.
+pub struct BlackMagicProbe {
+    reader: BufReader<Box<dyn Read + Send>>,
+    writer: BufWriter<Box<dyn Write + Send>>,
+    protocol: Option<WireProtocol>,
+    version: String,
+    remote_protocol: ProtocolVersion,
+    speed_khz: u32,
+    jtag_state: JtagDriverState,
+    probe_statistics: ProbeStatistics,
+    swd_settings: SwdSettings,
+    in_bits: BitVec<u8, Lsb0>,
+    swd_direction: SwdDirection,
+}
+
+impl core::fmt::Debug for BlackMagicProbe {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        write!(
+            f,
+            "Black Magic Probe {} with remote protocol {}",
+            self.version, self.remote_protocol
+        )
+    }
+}
+
+impl core::fmt::Display for BlackMagicProbeFactory {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        write!(f, "Black Magic Probe")
+    }
+}
+
+impl BlackMagicProbe {
+    fn new(
+        reader: Box<dyn Read + Send>,
+        writer: Box<dyn Write + Send>,
+    ) -> Result<Self, DebugProbeError> {
+        let mut reader = BufReader::new(reader);
+        let mut writer = BufWriter::new(writer);
+
+        let mut handshake_response = [0u8; 1024];
+        Self::send(
+            &mut writer,
+            &RemoteCommand::Handshake(&mut handshake_response),
+        )?;
+        let response_len = Self::recv(&mut reader, Some(&mut handshake_response), false)
+            .map_err(|e| {
+                tracing::error!("Unable to receive command: {:?}", e);
+                DebugProbeError::ProbeCouldNotBeCreated(ProbeCreationError::CouldNotOpen)
+            })?
+            .0;
+        let version =
+            String::from_utf8_lossy(&handshake_response[0..response_len as usize]).to_string();
+        tracing::info!("Probe version {}", version);
+
+        Self::send(&mut writer, &RemoteCommand::HighLevelCheck)?;
+        let remote_protocol = if let Ok(response) = Self::recv(&mut reader, None, false) {
+            match response.0 {
+                0 => ProtocolVersion::V0P,
+                1 => ProtocolVersion::V1,
+                2 => ProtocolVersion::V2,
+                3 => ProtocolVersion::V3,
+                4 => ProtocolVersion::V4,
+                _ => {
+                    return Err(DebugProbeError::ProbeCouldNotBeCreated(
+                        ProbeCreationError::Other("Unrecognized probe version response"),
```

Comment:
I would love some more output to the user here: Something along the lines `Unrecognized probe version response: Allowed values are in range [0, 4] but the received value was {0}`.

Reply:
How do I do that? `ProbeCreationError` requires a `&'static str`, which means dynamic strings cannot be formatted.

Reply:
I extended `RemoteError` and I now return that.

Reply:
Thanks!